apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts

```diff
@@ -0,0 +1,136 @@
+import { Component, OnInit, OnDestroy } from '@angular/core';
+import { Subscription } from 'rxjs';
+
+import { Todo, AddTodo } from '@learning-workspace/api-interfaces';
+import { ApiService } from '../../../../core/services';
+import { TodoActionEnum } from '../../enums';
+
+@Component({
+  selector: 'learning-workspace-todos-container',
+  templateUrl: './todos-container.component.html',
+  styleUrls: ['./todos-container.component.scss'],
+})
+export class TodosContainerComponent implements OnInit, OnDestroy {
+  public todos: Todo[] = [];
+  public selectedTodo: Todo;
+  public openAddEditForm = false;
+  public editMode = false;
+  public todoAction = '';
+
+  private subscription = new Subscription();
+
+  constructor(private readonly apiService: ApiService) {}
+
+  ngOnInit(): void {
+    this.getTodosSubscription();
+  }
+
+  ngOnDestroy(): void {
+    this.subscription.unsubscribe();
+  }
+
+  public addTodo(): void {
+    this.editMode = false;
+    this.openAddEditTodoForm();
+  }
+
+  public todoActionRun(action: string, id: string): void {
+    switch (action) {
+      case TodoActionEnum.check:
+        this.checkTodo(id);
+        break;
+
+      case TodoActionEnum.edit:
+        this.editTodo(id);
+        break;
+
+      case TodoActionEnum.remove:
+        this.removeTodo(id);
+        break;
+    }
+  }
+
+  public openAddEditTodoForm(): void {
+    this.openAddEditForm = !this.openAddEditForm;
+  }
+
+  public afterCloseAddEditTodoForm(todo: AddTodo | undefined): void {
+    this.openAddEditForm = false;
+
+    if (!todo?.title) {
+      return;
+    }
+
+    if (this.editMode) {
+      this.editTodoSubscription(this.selectedTodo.id, todo);
+      this.editMode = false;
+      return;
+    }
+
+    this.addTodoSubscription(todo);
+  }
+
+  public todoIdentify(_index: number, todo: Todo): string {
+    return todo.id;
+  }
+
+  private removeTodo(id: string): void {
+    this.removeTodoSubscription(id);
+  }
+
+  private editTodo(id: string): void {
+    this.editMode = true;
+    this.selectedTodo = this.todos.find((todo) => todo.id === id) as Todo;
+    this.openAddEditTodoForm();
+  }
+
+  private checkTodo(id: string): void {
+    this.checkTodoSubscription(id);
+  }
+
+  // TODO: change subscription to rxjs
+  private getTodosSubscription(): void {
+    const getTodosSubscription = this.apiService
```

Comment:
why do we need this const instead we can do like this this.subscription.add(your code)

Reply:
refactored it